src/lib/coda_base/transaction_union_tag.ml

```diff
@@ -3,71 +3,83 @@ open Fold_lib
 open Tuple_lib
 open Snark_params.Tick
 
-type t = Payment | Stake_delegation | Fee_transfer | Coinbase
+type t = Payment | Stake_delegation | Fee_transfer | Coinbase | Chain_voting
 [@@deriving enum, eq, sexp]
 
 let gen =
   Quickcheck.Generator.map (Int.gen_incl min max) ~f:(fun i ->
       Option.value_exn (of_enum i) )
 
-type var = Boolean.var * Boolean.var
-
-let to_bits = function
-  | Payment ->
-      (false, false)
-  | Stake_delegation ->
-      (true, false)
-  | Fee_transfer ->
-      (false, true)
-  | Coinbase ->
-      (true, true)
-
-let of_bits = function
-  | false, false ->
-      Payment
-  | true, false ->
-      Stake_delegation
-  | false, true ->
-      Fee_transfer
-  | true, true ->
-      Coinbase
-
-let%test_unit "to_bool of_bool inv" =
+type var = Boolean.var Triple.t
+
+let nth_bit x n = (x lsr n) land 1 = 1
+
+let list_to_triple_exn = function
+  | [x0; x1; x2] ->
+      (x0, x1, x2)
+  | _ ->
+      failwith "expected a list of length 3"
+
+let triple_to_list (x0, x1, x2) = [x0; x1; x2]
+
+let to_bits tag = List.init 3 ~f:(nth_bit (to_enum tag)) |> list_to_triple_exn
+
+let of_bits_exn bits =
+  Option.value_exn
+    ~error:(Error.of_string "unrecognized bits")
+    ( List.fold_right (triple_to_list bits) ~init:0 ~f:(fun b acc ->
+          (2 * acc) + Bool.to_int b )
+    |> of_enum )
+
+let%test_unit "to_bits of_bits inv" =
   let open Quickcheck in
-  test (Generator.tuple2 Bool.quickcheck_generator Bool.quickcheck_generator)
-    ~f:(fun b -> assert (b = to_bits (of_bits b)))
+  test
+    (Generator.tuple3 Bool.quickcheck_generator Bool.quickcheck_generator
+       (Generator.return false))
+    ~f:(fun b -> assert (b = to_bits (of_bits_exn b)))
 
 let typ =
-  Typ.transport Typ.(Boolean.typ * Boolean.typ) ~there:to_bits ~back:of_bits
+  Typ.transport
+    Typ.(tuple3 Boolean.typ Boolean.typ Boolean.typ)
+    ~there:to_bits ~back:of_bits_exn
 
-let fold (t : t) : bool Triple.t Fold.t =
+let fold (tag : t) : bool Triple.t Fold.t =
   { fold=
       (fun ~init ~f ->
-        let b0, b1 = to_bits t in
-        f init (b0, b1, false) ) }
+        let b0, b1, b2 = to_bits tag in
+        f init (b0, b1, b2) ) }
 
 let length_in_triples = 1
 
 module Checked = struct
   open Let_syntax
 
-  let constant t =
-    let x, y = to_bits t in
-    Boolean.(var_of_value x, var_of_value y)
+  let constant tag =
+    let b0, b1, b2 = to_bits tag in
+    Boolean.(var_of_value b0, var_of_value b1, var_of_value b2)
 
-  let to_triples ((x, y) : var) = [(x, y, Boolean.false_)]
+  let to_triples bits = [bits]
 
-  (* someday: Make these all cached *)
+  let is tag triple =
```

Comment:
Let `n = 3`. This is `2n + 2` constraints. You can do it in `2` constraints with

```ocaml
let is tag triple =
  let xs = triple_to_list (to_bits tag) in
  let ys = triple_to_list triple in
  let eq x y = if x then y else Boolean.not y in
  List.map2_exn xs ys ~f:eq |> Checked.all >>= Boolean.all
```

`n` is small so I know not a big deal, but I think might as well do the more efficient thing. We could also change `Boolean.equal` to do this automatically if one of the args is constant so you could just write the code as you did it originally.